torchchat/usages/server.py

```diff
@@ -4,38 +4,93 @@
 # This source code is licensed under the license found in the
 # LICENSE file in the root directory of this source tree.
 
+import atexit
 import json
 
 import logging
 
 logger = logging.getLogger(__name__)
 
+from contextlib import nullcontext
 from dataclasses import asdict
+from functools import partial
+from os import environ
 from typing import Dict, List, Union
 
 import torch
+import torch.distributed as dist
+import torch.multiprocessing as mp
+from concurrent import futures
 from flask import Flask, request, Response
 
 from torchchat.cli.builder import BuilderArgs, TokenizerArgs
+from torchchat.distributed.utils import setup_env
 from torchchat.generate import GeneratorArgs
 
 from torchchat.usages.openai_api import (
     CompletionRequest,
     get_model_info_list,
-    OpenAiApiGenerator,
+    create_openai_api_generator,
     retrieve_model_info,
 )
 
 OPENAI_API_VERSION = "v1"
 
 
+def run_worker(
+    args,
+    rank,
+    queue,
+    ):
+    """
+    This function creates and executes a generator 
+    """
+    gen = initialize_generator(args)
+    
+    while True:
+        try:
+            req = queue.get()
+        except KeyboardInterrupt:
+            break
+
+        if req == "stop":
+            break
+        
+        for _ in gen.chunked_completion(req):
+            pass
+
 def create_app(args):  # noqa: C901
     """
     Creates a flask app that can be used to serve the model as a chat API.
     """
     app = Flask(__name__)
 
-    gen: OpenAiApiGenerator = initialize_generator(args)
+    builder_args = BuilderArgs.from_args(args)
+    procs = []
+    if builder_args.distributed:
+        world_size = builder_args.tp * builder_args.pp
+        mp_context = mp.get_context('spawn')
+        queue = mp_context.Queue()
+    else:
+        world_size = 1
```

Comment:
Is the else case necessary? I don’t think these fields are used if not distributed